File: gui_application/src/api/lessons.rs
use chrono::Datelike;
use student::application::gateways::StudentLessonsGatewayError;
use student::domain::entities::{Clef, Lesson, Range, StudentLessons};

use crate::api::error_report::ErrorReportDto;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RangeDto {
    pub from: String,
    pub to: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ClefDto {
    G,
    C,
    F,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct DateDto {
    pub year: i32,
    pub month: u32,
    pub day: u32,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LessonDto {
    pub id: Option<String>,
    pub date: Option<DateDto>,
    pub phase: Option<RangeDto>,
    pub page: Option<RangeDto>,
    pub lesson: Option<RangeDto>,
    pub clef: Option<ClefDto>,
    pub description: Option<String>,
    pub instructor: Option<String>,
    pub method: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StudentLessonsDto {
    pub msa: Vec<LessonDto>,
    pub method: Vec<LessonDto>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum RetrieveStudentLessonsOutcome {
    Success { lessons: StudentLessonsDto },
    Failure { report: ErrorReportDto },
}

impl From<Result<StudentLessons, StudentLessonsGatewayError>> for RetrieveStudentLessonsOutcome {
    fn from(result: Result<StudentLessons, StudentLessonsGatewayError>) -> Self {
        match result {
            Ok(lessons) => Self::Success {
                lessons: lessons.into(),
            },
            Err(error) => Self::Failure {
                report: error.into(),
            },
        }
    }
}

impl From<StudentLessons> for StudentLessonsDto {
    fn from(lessons: StudentLessons) -> Self {
        Self {
            msa: lessons.msa.into_iter().map(LessonDto::from).collect(),
            method: lessons.method.into_iter().map(LessonDto::from).collect(),
        }
    }
}

impl From<Lesson> for LessonDto {
    fn from(lesson: Lesson) -> Self {
        Self {
            id: lesson.id,
            date: lesson.date.map(DateDto::from),
            phase: lesson.phase.map(RangeDto::from),
            page: lesson.page.map(RangeDto::from),
            lesson: lesson.lesson.map(RangeDto::from),
            clef: lesson.clef.map(ClefDto::from),
            description: lesson.description,
            instructor: lesson.instructor,
            method: lesson.method,
        }
    }
}

impl From<chrono::NaiveDate> for DateDto {
    fn from(date: chrono::NaiveDate) -> Self {
        Self {
            year: date.year(),
            month: date.month(),
            day: date.day(),
        }
    }
}

impl From<Range> for RangeDto {
    fn from(range: Range) -> Self {
        Self {
            from: range.from,
            to: range.to,
        }
    }
}

impl From<Clef> for ClefDto {
    fn from(clef: Clef) -> Self {
        match clef {
            Clef::G => Self::G,
            Clef::C => Self::C,
            Clef::F => Self::F,
        }
    }
}

File: gui_application/flutter/test/support/lessons.dart
import 'package:flutter_application/rust/api/error_report.dart';
import 'package:flutter_application/rust/api/lessons.dart';
import 'package:flutter_application/rust/api/progress.dart';

import 'pending.dart';

LessonDto lesson({
  String? id,
  (int, int, int)? date,
  (String, String)? phase,
  (String, String)? page,
  (String, String)? lesson,
  ClefDto? clef,
  String? description,
  String? instructor,
  String? method,
}) => LessonDto(
  id: id,
  date: date == null
      ? null
      : DateDto(year: date.$1, month: date.$2, day: date.$3),
  phase: _range(phase),
  page: _range(page),
  lesson: _range(lesson),
  clef: clef,
  description: description,
  instructor: instructor,
  method: method,
);

RangeDto? _range((String, String)? range) =>
    range == null ? null : RangeDto(from: range.$1, to: range.$2);

abstract final class Clefs {
  static const g = ClefDto.g;
}

StudentLessonsDto studentLessons({
  List<LessonDto> msa = const [],
  List<LessonDto> method = const [],
}) => StudentLessonsDto(msa: msa, method: method);

RetrieveStudentLessonsOutcomeDto lessonsRetrieved([
  StudentLessonsDto lessons = const StudentLessonsDto(msa: [], method: []),
]) => RetrieveStudentLessonsOutcomeDto.success(lessons: lessons);

Pending<RetrieveStudentLessonsOutcomeDto> pendingLessons() => Pending();

RetrieveStudentLessonsOutcomeDto lessonsFailed(ErrorReportDto report) =>
    RetrieveStudentLessonsOutcomeDto.failure(report: report);

abstract final class Levels {
  static const MusicianLevelDto candidate = MusicianLevelDto.candidate();
  static const MusicianLevelDto practice = MusicianLevelDto.practice();
  static const MusicianLevelDto youthService = MusicianLevelDto.youthService();
  static const MusicianLevelDto officialService =
      MusicianLevelDto.officialService();
  static const MusicianLevelDto officialized = MusicianLevelDto.officialized();

  static MusicianLevelDto unknown(String raw) =>
      MusicianLevelDto.unknown(raw: raw);
}

CheckpointStatusDto checkpointStatus(
  MusicianLevelDto level, {
  bool achieved = false,
  bool readyToAdvance = false,
}) => CheckpointStatusDto(
  level: level,
  achieved: achieved,
  readyToAdvance: readyToAdvance,
  requirement: const RequirementStatusDto(msaMet: true, methodMet: true),
);

ProgressAssessmentDto progressAssessment({
  List<CheckpointStatusDto> checkpoints = const [],
  double msaRelativePercent = 0,
  double methodRelativePercent = 0,
  double combinedPercent = 0,
  double overallCheckpointPercent = 0,
  MusicianLevelDto? nextLevel,
}) => ProgressAssessmentDto(
  checkpoints: checkpoints,
  msaRelativePercent: msaRelativePercent,
  methodRelativePercent: methodRelativePercent,
  combinedPercent: combinedPercent,
  overallCheckpointPercent: overallCheckpointPercent,
  nextLevel: nextLevel,
);

AssessStudentProgressOutcomeDto progressAssessed(
  ProgressAssessmentDto assessment,
) => AssessStudentProgressOutcomeDto.success(assessment: assessment);

AssessStudentProgressOutcomeDto noInstrumentAssigned() =>
    const AssessStudentProgressOutcomeDto.noInstrumentAssigned();

AssessStudentProgressOutcomeDto levelNotRecognized(String raw) =>
    AssessStudentProgressOutcomeDto.unknownLevel(rawLevel: raw);

AssessStudentProgressOutcomeDto notAMusician() =>
    const AssessStudentProgressOutcomeDto.notAMusician();

AssessStudentProgressOutcomeDto progressFailed(ErrorReportDto report) =>
    AssessStudentProgressOutcomeDto.failure(report: report);
